File: order.py
# Define the order class
class order:
    def __init__(self, id, time, price, quantity, is_sell, prev_order=None, next_order=None,\
        wrapper=None) -> None:
        # Order attributes
        # Private attributes
        self._id = id
        self._time = time 
        self._price = price
        self._quantity = quantity
        self._is_sell = is_sell
        
        # Neighbors
        self.prev_order = prev_order
        self.next_order = next_order
        self.wrapper = wrapper
        
    @property
    def id(self):
        return self._id
    
    @property
    def price(self):
        return self._price
    
    @property
    def is_sell(self):
        return self._is_sell
        
    @property
    def quantity(self):
        return self._quantity
        
    def append(self, new_order):
        self.next_order = new_order
        new_order.prev_order = self
        
    def pop(self):
        # Detach the order from the linked list
        if not self.prev_order and not self.next_order:
            # This is the only order in the LimitPrice node
            pass
        elif self.prev_order and not self.next_order:
            # This is the last order
            self.prev_order.next_order = None
            self.wrapper.order_tail = self.prev_order
            self.wrapper.total_size -= self.quantity
        elif not self.prev_order and self.next_order:
            # This is the first order
            self.prev_order.prev_order = None
            self.wrapper.order_head = self.next_order
            self.wrapper.total_size -= self.quantity
        else:
            # This is an order in the middle
            self.prev_order.next_order = self.next_order
            self.next_order.prev_order = self.prev_order
            self.wrapper.total_size -= self.quantity
            
    def update(self, new_order):
        assert self.price == new_order.price, "Order price doesn't match, maybe the order id is wrong"
        assert self.is_sell == new_order.is_sell, "Order type doesn't match, maybe the order id is wrong"
        self._time = new_order._time 
        if self.quantity == new_order.quantity:
            # Nothing different
            pass
        else:
            self.wrapper.total_size = self.wrapper.total_size - self.quantity + new_order.quantity
            self.quantity = new_order.quantity
        

class order_wrapper:
    """
    Wrapper of the doubly-linked list for order. It contains information about the head, tail, total_size, etc.
    """
    def __init__(self, new_order: order) -> None:
        self.LimitPrice = None
        self.order_head = new_order
        self.order_tail = new_order
        self.total_size = new_order.quantity
        self.price      = new_order.price
        
    def append(self, new_order: order):
        # The order belongs to the wrapper
        new_order.wrapper = self
        if not self.order_head:
            # First order
            self.order_head = new_order
            self.order_tail = new_order
            self.total_size = new_order.quantity
            self.price      = new_order.price
        else:
            assert self.price == new_order.price, "The new order's price doesn't match with the group"
            self.order_tail.append(new_order)
            self.order_tail = self.order_tail.next_order
            self.total_size += new_order.quantity
    

File: cpp/order.cpp
#include "order.h"
#include <cassert>

order::order(unsigned int _id, double _price, unsigned int _quantity, bool _is_sell)
{
    id = _id; price = _price; quantity = _quantity; is_sell = _is_sell;
}

void order::set_wrapper(order_summary* _wrapper)
{
    wrapper = _wrapper;
}

void order::set_quantity(unsigned int _new_quantity)
{
    quantity = _new_quantity;
}

void order::push_back(order* new_order)
{
    if (next_order){
        // There is an order following, we are iserting an order
        next_order->prev_order = new_order;
        new_order->next_order = next_order;
        new_order->prev_order = this;
        next_order = new_order;
        // Modify the wrapper
        wrapper->add_quantity(new_order->get_quantity());
    } else {
        // This is the end of the order chain
        next_order = new_order;
        new_order->prev_order = this;
        // Modify the wrapper
        wrapper->add_quantity(new_order->get_quantity());
        wrapper->order_tail = new_order;
    }
    // Set the wrapper of the new order
    new_order->set_wrapper(this->wrapper);
}

bool order::pop()
{
    // Pop the current node from the linked list
    if (!prev_order && !next_order){
        // This is the only order for the price
        wrapper->clear();
        return true;
    } else if (!prev_order && next_order) {
        // This is the first order for the price
        wrapper->add_quantity(-quantity);
        wrapper->order_head = next_order;
        next_order->prev_order = nullptr;
        return false;
    } else if (prev_order && !next_order) {
        // This is the last order for the price
        wrapper->add_quantity(-quantity);
        wrapper->order_tail = prev_order;
        prev_order->next_order = nullptr;
        return false;
    } else {
        // This is in the middle
        wrapper->add_quantity(-quantity);
        prev_order->next_order = next_order;
        next_order->prev_order = prev_order;
        return false;
    }

    // Reclaim the memory for the order
    delete this;
}

void order::update(order* new_order)
{
    // Update the order information
    assert(new_order->get_id()==id && "The id doesn't match!");
    // Push back to the tail of the linked list then pop the current
    wrapper->order_tail->push_back(new_order);
    pop();
}

void order_summary::add_quantity(int delta_quant)
{
    total_quantity += delta_quant;
}

order_summary::order_summary(order* new_order)
{
    price = new_order->get_price();
    total_quantity = new_order->get_quantity();
    order_head = new_order;
    order_tail = new_order;
    new_order->set_wrapper(this);
}

order_summary::~order_summary()
{
    order* tmp = order_head;
    while (tmp){
        order* tp = tmp->next_order;
        delete tmp;
        tmp = tp;
    }
}
